File: rust/transports/local/src/transport.rs
//! Rust counterpart of `c++/transports/local/include/transports/local/transport.h`.
//!
//! The primary entities mirror the C++ local transport family:
//! - `ChildTransport`: parent zone -> child zone
//! - `ParentTransport`: child zone -> parent zone
//! - `ChildZone`: helper that binds a child service to that transport pair

use canopy_rpc::internal::error_codes;
use canopy_rpc::{
    AddRefParams, GetNewZoneIdParams, IMarshaller, NewZoneIdResult, Object, ObjectReleasedParams,
    PostParams, ReleaseParams, RemoteObject, SendParams, SendResult, ServiceRuntime,
    StandardResult, Transport, TransportDownParams, TransportStatus, TryCastParams,
};
use std::sync::{Arc, Mutex, Weak};

#[derive(Default)]
struct LocalPeerMarshaller {
    peer: Mutex<Weak<Transport>>,
}

impl LocalPeerMarshaller {
    fn set_peer(&self, peer: &Arc<Transport>) {
        *self.peer.lock().expect("local peer mutex poisoned") = Arc::downgrade(peer);
    }

    fn peer(&self) -> Option<Arc<Transport>> {
        self.peer
            .lock()
            .expect("local peer mutex poisoned")
            .upgrade()
    }
}

fn weak_service(service: &Arc<dyn ServiceRuntime>) -> Weak<dyn ServiceRuntime> {
    Arc::downgrade(service)
}

impl IMarshaller for LocalPeerMarshaller {
    fn send(&self, params: SendParams) -> SendResult {
        let Some(peer) = self.peer() else {
            return SendResult::new(error_codes::ZONE_NOT_FOUND(), vec![], vec![]);
        };
        peer.service()
            .map(|service| service.send(params))
            .unwrap_or_else(|| SendResult::new(error_codes::ZONE_NOT_FOUND(), vec![], vec![]))
    }

    fn post(&self, params: PostParams) {
        if let Some(service) = self.peer().and_then(|peer| peer.service()) {
            service.post(params);
        }
    }

    fn try_cast(&self, params: TryCastParams) -> StandardResult {
        self.peer()
            .and_then(|peer| peer.service())
            .map(|service| service.try_cast(params))
            .unwrap_or_else(|| StandardResult::new(error_codes::ZONE_NOT_FOUND(), vec![]))
    }

    fn add_ref(&self, params: AddRefParams) -> StandardResult {
        let Some(peer) = self.peer() else {
            return StandardResult::new(error_codes::ZONE_NOT_FOUND(), vec![]);
        };
        peer.inbound_add_ref(params)
    }

    fn release(&self, params: ReleaseParams) -> StandardResult {
        self.peer()
            .map(|peer| peer.inbound_release(params))
            .unwrap_or_else(|| StandardResult::new(error_codes::ZONE_NOT_FOUND(), vec![]))
    }

    fn object_released(&self, params: ObjectReleasedParams) {
        if let Some(service) = self.peer().and_then(|peer| peer.service()) {
            service.object_released(params);
        }
    }

    fn transport_down(&self, params: TransportDownParams) {
        if let Some(service) = self.peer().and_then(|peer| peer.service()) {
            service.transport_down_from_params(params);
        }
    }

    fn get_new_zone_id(&self, params: GetNewZoneIdParams) -> NewZoneIdResult {
        self.peer()
            .and_then(|peer| peer.service())
            .map(|service| service.get_new_zone_id(params))
            .unwrap_or_else(|| {
                NewZoneIdResult::new(error_codes::ZONE_NOT_FOUND(), Default::default(), vec![])
            })
    }
}

#[derive(Clone)]
pub struct ChildTransport {
    transport: Arc<Transport>,
}

impl ChildTransport {
    pub fn transport(&self) -> Arc<Transport> {
        self.transport.clone()
    }
}

#[derive(Clone)]
pub struct ParentTransport {
    transport: Arc<Transport>,
}

impl ParentTransport {
    pub fn transport(&self) -> Arc<Transport> {
        self.transport.clone()
    }
}

#[derive(Clone)]
pub struct TransportPair {
    child_transport: ChildTransport,
    parent_transport: ParentTransport,
}

#[derive(Clone)]
pub struct ChildZone {
    service: Arc<canopy_rpc::ChildService>,
    transports: TransportPair,
}

#[derive(Clone)]
pub struct BoundChildZone {
    zone: ChildZone,
    root_object: RemoteObject,
}

impl ChildZone {
    pub fn create(
        name: impl Into<String>,
        parent_service: Arc<dyn ServiceRuntime>,
        child_name: impl Into<String>,
        child_zone_id: canopy_rpc::Zone,
    ) -> Self {
        let service = canopy_rpc::ChildService::new_shared(
            child_name,
            child_zone_id,
            parent_service.zone_id(),
        );
        let transports = TransportPair::connect(name, parent_service, service.clone());
        service.set_parent_transport(transports.parent_transport().transport());
        Self {
            service,
            transports,
        }
    }

    pub fn service(&self) -> &Arc<canopy_rpc::ChildService> {
        &self.service
    }

    pub fn transports(&self) -> &TransportPair {
        &self.transports
    }

    pub fn child_transport(&self) -> &ChildTransport {
        self.transports.child_transport()
    }

    pub fn parent_transport(&self) -> &ParentTransport {
        self.transports.parent_transport()
    }

    pub fn bind_root_rpc_object(
        self,
        parent_service: Arc<dyn ServiceRuntime>,
        object_id: Object,
        root_object: Arc<dyn canopy_rpc::RpcObject>,
        pointer_kind: canopy_rpc::InterfacePointerKind,
    ) -> Result<BoundChildZone, i32> {
        self.service()
            .register_rpc_object(object_id, root_object)
            .map_err(|error_code| error_code)?;
        let remote_object = self
            .service()
            .zone_id()
            .with_object(object_id)
            .map_err(|_| error_codes::INVALID_DATA())?;
        let add_ref = self.service().add_ref(AddRefParams {
            protocol_version: canopy_rpc::DefaultValues::VERSION_3 as u64,
            remote_object_id: remote_object.clone(),
            caller_zone_id: parent_service.zone_id(),
            requesting_zone_id: self.service().zone_id(),
            build_out_param_channel: canopy_rpc::add_ref_options_for_pointer_kind(pointer_kind),
            in_back_channel: vec![],
        });
        if error_codes::is_error(add_ref.error_code) {
            return Err(add_ref.error_code);
        }
        Ok(BoundChildZone {
            zone: self,
            root_object: remote_object,
        })
    }
}

impl BoundChildZone {
    pub fn zone(&self) -> &ChildZone {
        &self.zone
    }

    pub fn root_object(&self) -> &RemoteObject {
        &self.root_object
    }
}

impl TransportPair {
    pub fn connect(
        name: impl Into<String>,
        parent_service: Arc<dyn ServiceRuntime>,
        child_service: Arc<dyn ServiceRuntime>,
    ) -> Self {
        let name = name.into();
        let parent_to_child_peer = Arc::new(LocalPeerMarshaller::default());
        let child_to_parent_peer = Arc::new(LocalPeerMarshaller::default());

        let parent_to_child = Transport::new(
            format!("{name}:parent-to-child"),
            parent_service.zone_id(),
            child_service.zone_id(),
            weak_service(&parent_service),
            parent_to_child_peer.clone(),
        );
        let child_to_parent = Transport::new(
            format!("{name}:child-to-parent"),
            child_service.zone_id(),
            parent_service.zone_id(),
            weak_service(&child_service),
            child_to_parent_peer.clone(),
        );

        parent_to_child_peer.set_peer(&child_to_parent);
        child_to_parent_peer.set_peer(&parent_to_child);
        parent_to_child.set_status(TransportStatus::Connected);
        child_to_parent.set_status(TransportStatus::Connected);
        parent_service.add_transport(child_service.zone_id(), parent_to_child.clone());
        child_service.add_transport(parent_service.zone_id(), child_to_parent.clone());

        Self {
            child_transport: ChildTransport {
                transport: parent_to_child,
            },
            parent_transport: ParentTransport {
                transport: child_to_parent,
            },
        }
    }

    pub fn child_transport(&self) -> &ChildTransport {
        &self.child_transport
    }

    pub fn parent_transport(&self) -> &ParentTransport {
        &self.parent_transport
    }
}

#[doc(hidden)]
pub type LocalChildTransport = ChildTransport;
#[doc(hidden)]
pub type LocalParentTransport = ParentTransport;
#[doc(hidden)]
pub type LocalTransportPair = TransportPair;
#[doc(hidden)]
pub type LocalChildZone = ChildZone;
#[doc(hidden)]
pub type LocalBoundChildZone = BoundChildZone;

#[cfg(test)]
mod tests {
    use super::*;
    use canopy_rpc::{
        AddressType, ChildService, DefaultValues, Encoding, InterfaceOrdinal, Method, Object,
        RemoteObject, RootService, Zone, ZoneAddress, ZoneAddressArgs,
    };
    use std::sync::Mutex;

    #[derive(Debug)]
    struct RecordingTarget {
        calls: Mutex<u32>,
    }

    impl canopy_rpc::CastingInterface for RecordingTarget {
        fn __rpc_query_interface(&self, _interface_id: InterfaceOrdinal) -> bool {
            true
        }

        fn __rpc_call(&self, _params: SendParams) -> SendResult {
            *self.calls.lock().expect("calls mutex poisoned") += 1;
            SendResult::new(error_codes::OK(), vec![1, 2, 3], vec![])
        }
    }

    impl canopy_rpc::GeneratedRustInterface for RecordingTarget {
        fn interface_name() -> &'static str {
            "RecordingTarget"
        }

        fn get_id(_rpc_version: u64) -> u64 {
            77
        }

        fn binding_metadata() -> &'static [canopy_rpc::GeneratedMethodBindingDescriptor] {
            &[]
        }
    }

    fn zone(subnet: u32) -> Zone {
        Zone::new(
            ZoneAddress::create(ZoneAddressArgs::new(
                DefaultValues::VERSION_3,
                AddressType::Local,
                0,
                vec![],
                DefaultValues::DEFAULT_SUBNET_SIZE_BITS,
                subnet.into(),
                DefaultValues::DEFAULT_OBJECT_ID_SIZE_BITS,
                0,
                vec![],
            ))
            .expect("zone address should be valid"),
        )
    }

    #[test]
    fn local_pair_forwards_parent_to_child_send() {
        let parent = RootService::new_shared("parent", zone(1));
        let child = ChildService::new_shared("child", zone(2), parent.zone_id());
        let _pair = TransportPair::connect("test", parent.clone(), child.clone());
        let target = Arc::new(RecordingTarget {
            calls: Mutex::new(0),
        });
        child
            .register_local_object(Object::new(9), target.clone())
            .expect("register child target");
        let remote = child.zone_id().with_object(Object::new(9)).expect("remote");
        let result = parent
            .get_transport(&child.zone_id())
            .expect("parent transport")
            .send(SendParams {
                protocol_version: DefaultValues::VERSION_3 as u64,
                encoding_type: Encoding::ProtocolBuffers,
                tag: 0,
                caller_zone_id: parent.zone_id(),
                remote_object_id: remote,
                interface_id: InterfaceOrdinal::new(77),
                method_id: Method::new(1),
                in_data: vec![],
                in_back_channel: vec![],
            });
        assert_eq!(result.error_code, error_codes::OK());
        assert_eq!(*target.calls.lock().expect("calls mutex poisoned"), 1);
    }

    #[test]
    fn local_pair_forwards_child_to_parent_add_ref() {
        let parent = RootService::new_shared("parent", zone(10));
        let child = ChildService::new_shared("child", zone(11), parent.zone_id());
        let _pair = TransportPair::connect("test", parent.clone(), child.clone());
        let target = Arc::new(RecordingTarget {
            calls: Mutex::new(0),
        });
        parent
            .register_local_object(Object::new(3), target)
            .expect("register parent target");
        let remote = RemoteObject::new(parent.zone_id().get_address().clone())
            .with_object(Object::new(3))
            .expect("remote");
        let result = child
            .get_transport(&parent.zone_id())
            .expect("child transport")
            .add_ref(AddRefParams {
                protocol_version: DefaultValues::VERSION_3 as u64,
                remote_object_id: remote,
                caller_zone_id: child.zone_id(),
                requesting_zone_id: child.zone_id(),
                build_out_param_channel: canopy_rpc::AddRefOptions::NORMAL,
                in_back_channel: vec![],
            });
        assert_eq!(result.error_code, error_codes::OK());
    }
}

File: c++/transports/local/src/transport.cpp
/*
 *   Copyright (c) 2026 Edward Boggis-Rolfe
 *   All rights reserved.
 */

#include <transports/local/transport.h>
#include <rpc/rpc.h>

namespace rpc::local
{
    parent_transport::parent_transport(
        std::string name,
        std::shared_ptr<rpc::service> service,
        std::shared_ptr<child_transport> parent)
        : rpc::transport(
              name,
              service)
        , parent_(parent)
    {
        // Local transports are always immediately available (in-process)
        set_status(rpc::transport_status::CONNECTED);
    }

    parent_transport::parent_transport(
        std::string name,
        std::shared_ptr<child_transport> parent)
        : rpc::transport(
              name,
              parent->get_adjacent_zone_id())
        , parent_(parent)
    {
        set_adjacent_zone_id(parent->get_zone_id());
        // Local transports are always immediately available (in-process)
        set_status(rpc::transport_status::CONNECTED);
    }

    void parent_transport::set_status(rpc::transport_status status)
    {
        // Call base class to update status
        rpc::transport::set_status(status);

        // If disconnecting, notify parent zone to break circular reference
        if (status == rpc::transport_status::DISCONNECTED)
        {
            auto parent = parent_.get_nullable();
            if (parent)
            {
                // Notify parent zone's child_transport to break its child_ reference
                parent->on_child_disconnected();
            }
            // Break our reference to parent
            parent_.reset();
        }
    }

    // Outbound i_marshaller interface - sends from child to parent
    CORO_TASK(send_result)
    parent_transport::outbound_send(send_params params)
    {
        auto parent = parent_.get_nullable();
        if (!parent)
        {
            RPC_ERROR("parent_transport::outbound_send: parent is NULL!");
            CO_RETURN send_result{rpc::error::ZONE_NOT_FOUND(), {}, {}};
        }

        // Forward to parent transport's inbound handler
        CO_RETURN CO_AWAIT parent->inbound_send(std::move(params));
    }

    CORO_TASK(void)
    parent_transport::outbound_post(post_params params)
    {
        auto parent = parent_.get_nullable();
        if (!parent)
        {
            RPC_ERROR("parent_transport::outbound_post: parent is NULL!");
            CO_RETURN;
        }

        CO_AWAIT parent->inbound_post(std::move(params));
    }

    CORO_TASK(standard_result)
    parent_transport::outbound_try_cast(try_cast_params params)
    {
        auto parent = parent_.get_nullable();
        if (!parent)
        {
            RPC_ERROR("parent_transport::outbound_try_cast: parent is NULL!");
            CO_RETURN standard_result{rpc::error::ZONE_NOT_FOUND(), {}};
        }

        CO_RETURN CO_AWAIT parent->inbound_try_cast(std::move(params));
    }

    CORO_TASK(standard_result)
    parent_transport::outbound_add_ref(add_ref_params params)
    {
        RPC_DEBUG(
            "parent_transport::outbound_add_ref: my_zone={}, adjacent_zone={}, destination_zone={}, caller_zone={}",
            get_zone_id().get_subnet(),
            get_adjacent_zone_id().get_subnet(),
            params.remote_object_id.get_subnet(),
            params.caller_zone_id.get_subnet());

        auto parent = parent_.get_nullable();
        if (!parent)
        {
            RPC_ERROR("parent_transport::outbound_add_ref: parent is NULL!");
            CO_RETURN standard_result{rpc::error::ZONE_NOT_FOUND(), {}};
        }

        RPC_DEBUG(
            "parent_transport::outbound_add_ref: Calling parent->inbound_add_ref for zone {}",
            params.remote_object_id.get_subnet());
        CO_RETURN CO_AWAIT parent->inbound_add_ref(std::move(params));
    }

    CORO_TASK(standard_result)
    parent_transport::outbound_release(release_params params)
    {
        auto parent = parent_.get_nullable();
        if (!parent)
        {
            RPC_ERROR("parent_transport::outbound_release: parent is NULL!");
            CO_RETURN standard_result{rpc::error::ZONE_NOT_FOUND(), {}};
        }

        CO_RETURN CO_AWAIT parent->inbound_release(std::move(params));
    }

    CORO_TASK(void)
    parent_transport::outbound_object_released(object_released_params params)
    {
        auto parent = parent_.get_nullable();
        if (!parent)
        {
            RPC_ERROR("parent_transport::outbound_object_released: parent is NULL!");
            CO_RETURN;
        }

        CO_AWAIT parent->inbound_object_released(std::move(params));
    }

    CORO_TASK(void)
    parent_transport::outbound_transport_down(transport_down_params params)
    {
        auto parent = parent_.get_nullable();
        if (!parent)
        {
            RPC_ERROR("parent_transport::outbound_transport_down: parent is NULL!");
            CO_RETURN;
        }

        CO_AWAIT parent->inbound_transport_down(std::move(params));
    }

    CORO_TASK(new_zone_id_result)
    parent_transport::outbound_get_new_zone_id(get_new_zone_id_params params)
    {
        auto parent = parent_.get_nullable();
        if (!parent)
        {
            RPC_ERROR("parent_transport::outbound_get_new_zone_id: parent is NULL!");
            CO_RETURN new_zone_id_result{rpc::error::ZONE_NOT_FOUND(), {}, {}};
        }
        CO_RETURN CO_AWAIT parent->get_new_zone_id(std::move(params));
    }

    // Transport from parent zone to child zone
    // Used by parent to communicate with child

    void child_transport::on_child_disconnected()
    {
        // Break circular reference when child zone disconnects
        // Safe because stack-based shared_ptr in outbound_* methods keeps parent_transport alive
        child_.reset();
    }

    // Outbound i_marshaller interface - sends from parent to child
    CORO_TASK(send_result)
    child_transport::outbound_send(send_params params)
    {
        auto child = child_.get_nullable();
        if (!child)
        {
            CO_RETURN send_result{rpc::error::ZONE_NOT_FOUND(), {}, {}};
        }

        // Forward to child service's inbound handler via its transport
        // Since child_service IS an i_marshaller, we can call directly
        CO_RETURN CO_AWAIT child->inbound_send(std::move(params));
    }

    CORO_TASK(void)
    child_transport::outbound_post(post_params params)
    {
        auto child = child_.get_nullable();
        if (!child)
        {
            CO_RETURN;
        }

        CO_AWAIT child->inbound_post(std::move(params));
    }

    CORO_TASK(standard_result)
    child_transport::outbound_try_cast(try_cast_params params)
    {
        auto child = child_.get_nullable();
        if (!child)
        {
            CO_RETURN standard_result{rpc::error::ZONE_NOT_FOUND(), {}};
        }

        CO_RETURN CO_AWAIT child->inbound_try_cast(std::move(params));
    }

    CORO_TASK(standard_result)
    child_transport::outbound_add_ref(add_ref_params params)
    {
        auto child = child_.get_nullable();
        if (!child)
        {
            CO_RETURN standard_result{rpc::error::ZONE_NOT_FOUND(), {}};
        }

        CO_RETURN CO_AWAIT child->inbound_add_ref(std::move(params));
    }

    CORO_TASK(standard_result)
    child_transport::outbound_release(release_params params)
    {
        auto child = child_.get_nullable();
        if (!child)
        {
            CO_RETURN standard_result{rpc::error::ZONE_NOT_FOUND(), {}};
        }

        CO_RETURN CO_AWAIT child->inbound_release(std::move(params));
    }

    CORO_TASK(void)
    child_transport::outbound_object_released(object_released_params params)
    {
        auto child = child_.get_nullable();
        if (!child)
        {
            CO_RETURN;
        }

        CO_AWAIT child->inbound_object_released(std::move(params));
    }

    CORO_TASK(void)
    child_transport::outbound_transport_down(transport_down_params params)
    {
        auto child = child_.get_nullable();
        if (!child)
        {
            CO_RETURN;
        }

        CO_AWAIT child->inbound_transport_down(std::move(params));
    }
}
